janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/rest/RestElasticSearchClient.java

```diff
@@ -416,17 +452,107 @@ public void bulkRequest(List<ElasticSearchMutation> requests, String ingestPipel
+                    chunkSerializedTotal += requestSerializedSize;
+                    serializedRequests.add(requestIterator.next());
+                } else if (requestSerializedSize > bulkChunkSerializedLimitBytes) {
+                    //we've encountered an element we cannot send to Elasticsearch given the configured limit
+                    throw new IllegalArgumentException(String.format(
+                        "Bulk request item is larger than permitted chunk limit. Limit is %s. Serialized item size was %s",
+                        bulkChunkSerializedLimitBytes, requestSerializedSize));
+                } else {
+                    //Adding this element would exceed the limit, so return the chunk
+                    return serializedRequests;
+                }
+            }
+            //All remaining requests fit in this chunk
+            return serializedRequests;
+        }
+    }
+
+    @Override
+    public void bulkRequest(final List<ElasticSearchMutation> requests, String ingestPipeline) throws IOException {
+        BulkRequestChunker bulkRequestChunker = new BulkRequestChunker(requests);
+        while (bulkRequestChunker.hasNext()) {
+            List<RequestBytes> bulkRequestChunk = bulkRequestChunker.next();
+            int retryCount = 0;
+            while (true) {
+                final Pair<String, byte[]> bulkRequestInput = buildBulkRequestInput(bulkRequestChunk, ingestPipeline);
+                final Response response = performRequest(REQUEST_TYPE_POST, bulkRequestInput.getValue0(), bulkRequestInput.getValue1());
+                try (final InputStream inputStream = response.getEntity().getContent()) {
+                    final RestBulkResponse bulkResponse = mapper.readValue(inputStream, RestBulkResponse.class);
+                    List<Triplet<Object, Integer, RequestBytes>> bulkItemsThatFailed = pairErrorsWithSubmittedMutation(bulkResponse.getItems(), bulkRequestChunk);
+                    if (!bulkItemsThatFailed.isEmpty()) {
+                        //Only retry the bulk request if *all* the bulk response item error codes are retry error codes
+                        final Set<Integer> errorCodes = bulkItemsThatFailed.stream().map(Triplet::getValue1).collect(Collectors.toSet());
+                        if (retryCount < retryAttemptLimit && retryOnErrorCodes.containsAll(errorCodes)) {
+                            //Build up the next request batch, of only the failed mutations
+                            bulkRequestChunk = bulkItemsThatFailed.stream().map(Triplet::getValue2).collect(Collectors.toList());
+                            performRetryWait(retryCount);
+                            retryCount++;
+                        } else {
+                            final List<Object> errorItems = bulkItemsThatFailed.stream().map(Triplet::getValue0).collect(Collectors.toList());
+                            errorItems.forEach(error -> log.error("Failed to execute ES query: {}", error));
```

Comment:
You don't need `{}` here
```suggestion
                            errorItems.forEach(error -> log.error("Failed to execute ES query:", error));
```

Reply:
Oh damn sorry for that. I thought it was an exception